staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go

```diff
@@ -1239,23 +1239,19 @@ func (e *Store) Watch(ctx context.Context, options *metainternalversion.ListOpti
 
 // WatchPredicate starts a watch for the items that matches.
 func (e *Store) WatchPredicate(ctx context.Context, p storage.SelectionPredicate, resourceVersion string) (watch.Interface, error) {
-	storageOpts := storage.ListOptions{ResourceVersion: resourceVersion, Predicate: p}
+	storageOpts := storage.ListOptions{ResourceVersion: resourceVersion, Predicate: p, Recursive: true}
+
+	key := e.KeyRootFunc(ctx)
 	if name, ok := p.MatchesSingle(); ok {
```

Comment:
If `p.MatchesSingle()` is true but `e.KeyFunc(ctx, name)` returns an error, should `storageOpts.Recursive` still be set to `false`?

Reply:
Yeah - this is consistent with what we had before, where we were calling WatchList with KeyRootFunc below.